File: scripts/sync_product_highlights.js
const fs = require("fs");
const path = require("path");

const TEMPLATES_DIR = path.join(__dirname, "..", "templates");
const BUNDLES_FILE = path.join(TEMPLATES_DIR, "product.bundles.json");
const FLAG_BLOCK_ID = "product_flag_qn3MLq";
const PAYMENT_BLOCK_ID = "custom_liquid_kj8teT";
const OLD_FLAG_BLOCK_ID = "product_flag_T9zT3n";
const MAIN_SECTION = "main";

const MARKET_ALIASES = {
  denmark: "danmark",
  norway: "norge",
  sweden: "svergie",
};

const NORDIC_FLAG_SVGS = {
  danmark:
    '<svg xmlns="http://www.w3.org/2000/svg" viewbox="0 0 370 280" width="370" height="280">\n  <rect width="370" height="280" fill="#C8102E"></rect>\n  <rect x="120" width="40" height="280" fill="#FFFFFF"></rect>\n  <rect y="120" width="370" height="40" fill="#FFFFFF"></rect>\n</svg>',
  denmark:
    '<svg xmlns="http://www.w3.org/2000/svg" viewbox="0 0 370 280" width="370" height="280">\n  <rect width="370" height="280" fill="#C8102E"></rect>\n  <rect x="120" width="40" height="280" fill="#FFFFFF"></rect>\n  <rect y="120" width="370" height="40" fill="#FFFFFF"></rect>\n</svg>',
  norge:
    '<svg xmlns="http://www.w3.org/2000/svg" viewbox="0 0 220 160" width="220" height="160">\n  <rect width="220" height="160" fill="#BA0C2F"></rect>\n  <rect x="50" width="40" height="160" fill="#FFFFFF"></rect>\n  <rect y="60" width="220" height="40" fill="#FFFFFF"></rect>\n  <rect x="60" width="20" height="160" fill="#00205B"></rect>\n  <rect y="70" width="220" height="20" fill="#00205B"></rect>\n</svg>',
  norway:
    '<svg xmlns="http://www.w3.org/2000/svg" viewbox="0 0 220 160" width="220" height="160">\n  <rect width="220" height="160" fill="#BA0C2F"></rect>\n  <rect x="50" width="40" height="160" fill="#FFFFFF"></rect>\n  <rect y="60" width="220" height="40" fill="#FFFFFF"></rect>\n  <rect x="60" width="20" height="160" fill="#00205B"></rect>\n  <rect y="70" width="220" height="20" fill="#00205B"></rect>\n</svg>',
};

const NORDIC_MARKETS = Object.keys(NORDIC_FLAG_SVGS);

const AUTO_HEADER = `/*
 * ------------------------------------------------------------
 * IMPORTANT: The contents of this file are auto-generated.
 *
 * This file may be updated by the Shopify admin theme editor
 * or related systems. Please exercise caution as any changes
 * made to this file may be overwritten.
 * ------------------------------------------------------------
 */
`;

function getNordicFlagSvg(market) {
  if (NORDIC_FLAG_SVGS[market]) return NORDIC_FLAG_SVGS[market];
  const alias = MARKET_ALIASES[market];
  return alias ? NORDIC_FLAG_SVGS[alias] : undefined;
}

function isNordicMarket(market) {
  return Boolean(getNordicFlagSvg(market));
}

function stripJsonComments(text) {
  return text.replace(/\/\*[\s\S]*?\*\//g, "");
}

function loadJson(filePath) {
  return JSON.parse(stripJsonComments(fs.readFileSync(filePath, "utf8")));
}

function saveJson(filePath, data) {
  const isNew = !fs.existsSync(filePath);
  const original = isNew ? AUTO_HEADER : fs.readFileSync(filePath, "utf8");
  const headerMatch = original.match(/^\/\*[\s\S]*?\*\/\s*/);
  const header = headerMatch ? headerMatch[0] : AUTO_HEADER;
  fs.writeFileSync(filePath, header + JSON.stringify(data, null, 2) + "\n", "utf8");
}

function deepMerge(base, override) {
  const result = structuredClone(base);
  for (const [key, value] of Object.entries(override)) {
    if (
      Object.prototype.hasOwnProperty.call(result, key) &&
      typeof result[key] === "object" &&
      result[key] !== null &&
      typeof value === "object" &&
      value !== null &&
      !Array.isArray(value)
    ) {
      result[key] = deepMerge(result[key], value);
    } else {
      result[key] = structuredClone(value);
    }
  }
  return result;
}

function findPaymentBlockIds(blocks) {
  const ids = [];
  for (const [blockId, block] of Object.entries(blocks)) {
    if (block.type === "custom_liquid" && block.name === "Payment Methods") {
      ids.push(blockId);
    }
    if (
      block.type === "section_vso_theme_sections_payment_icons" ||
      (typeof block.type === "string" && block.type.includes("paymentIcons")) ||
      (blockId.includes("payment_icons") && blockId !== PAYMENT_BLOCK_ID)
    ) {
      ids.push(blockId);
    }
  }
  return ids;
}

function cleanBlockOrder(blockOrder, blocks) {
  return blockOrder.filter((id) => Object.prototype.hasOwnProperty.call(blocks, id));
}

function fixBlockOrder(blockOrder, paymentIdsToRemove) {
  const remove = new Set([OLD_FLAG_BLOCK_ID, ...paymentIdsToRemove]);
  const order = blockOrder.filter((id) => !remove.has(id));

  if (!order.includes("buy_buttons")) {
    return order;
  }

  const buyIdx = order.indexOf("buy_buttons");
  const insertAt = buyIdx + 1;

  const flagIdx = order.indexOf(FLAG_BLOCK_ID);
  if (flagIdx !== -1) order.splice(flagIdx, 1);
  const paymentIdx = order.indexOf(PAYMENT_BLOCK_ID);
  if (paymentIdx !== -1) order.splice(paymentIdx, 1);

  order.splice(insertAt, 0, FLAG_BLOCK_ID, PAYMENT_BLOCK_ID);
  return order;
}

function loadBundlesReference() {
  const bundles = loadJson(BUNDLES_FILE);
  const main = bundles.sections[MAIN_SECTION];
  return {
    flagBlock: structuredClone(main.blocks[FLAG_BLOCK_ID]),
    paymentBlock: structuredClone(main.blocks[PAYMENT_BLOCK_ID]),
  };
}

function loadMarketFlagOverrides() {
  const overrides = {};
  const files = fs
    .readdirSync(TEMPLATES_DIR)
    .filter((name) => name.startsWith("product.bundles.context.") && name.endsWith(".json"))
    .sort();

  for (const name of files) {
    const data = loadJson(path.join(TEMPLATES_DIR, name));
    const market = data.context?.market;
    if (!market) continue;

    const block = data.sections?.[MAIN_SECTION]?.blocks?.[FLAG_BLOCK_ID];
    if (!block) continue;

    overrides[market] = structuredClone(block);
    const alias = MARKET_ALIASES[market];
    if (alias && !overrides[alias]) {
      overrides[alias] = structuredClone(block);
    }
  }

  for (const market of NORDIC_MARKETS) {
    overrides[market] = {
      disabled: false,
      settings: {
        icon_1_svg: NORDIC_FLAG_SVGS[market],
      },
    };
  }

  return overrides;
}

function processBaseTemplate(filePath, flagBlock, paymentBlock) {
  if (path.basename(filePath) === path.basename(BUNDLES_FILE)) {
    return false;
  }

  const data = loadJson(filePath);
  const main = data.sections?.[MAIN_SECTION];
  if (!main) return false;

  const blocks = main.blocks || (main.blocks = {});
  let changed = false;

  if (blocks[OLD_FLAG_BLOCK_ID]) {
    delete blocks[OLD_FLAG_BLOCK_ID];
    changed = true;
  }

  const paymentIdsToRemove = new Set(findPaymentBlockIds(blocks));
  paymentIdsToRemove.delete(PAYMENT_BLOCK_ID);

  for (const paymentId of paymentIdsToRemove) {
    if (blocks[paymentId]) {
      delete blocks[paymentId];
      changed = true;
    }
  }

  if (JSON.stringify(blocks[FLAG_BLOCK_ID]) !== JSON.stringify(flagBlock)) {
    blocks[FLAG_BLOCK_ID] = structuredClone(flagBlock);
    changed = true;
  }

  if (JSON.stringify(blocks[PAYMENT_BLOCK_ID]) !== JSON.stringify(paymentBlock)) {
    blocks[PAYMENT_BLOCK_ID] = structuredClone(paymentBlock);
    changed = true;
  }

  if (main.block_order) {
    const newOrder = fixBlockOrder(cleanBlockOrder(main.block_order, blocks), paymentIdsToRemove);
    if (JSON.stringify(newOrder) !== JSON.stringify(main.block_order)) {
      main.block_order = newOrder;
      changed = true;
    }
  }

  if (changed) saveJson(filePath, data);
  return changed;
}

function processContextTemplate(filePath, marketOverrides) {
  if (path.basename(filePath).includes("product.bundles.context.")) {
    return false;
  }

  const data = loadJson(filePath);
  const market = data.context?.market;
  if (!market) return false;

  const main = data.sections?.[MAIN_SECTION];
  if (!main) return false;

  const blocks = main.blocks || (main.blocks = {});
  let changed = false;

  if (isNordicMarket(market)) {
    const contextData = { sections: { [MAIN_SECTION]: main }, context: { market } };
    if (applyNordicFlagToContextData(contextData, market)) {
      changed = true;
    }
  } else {
    let flagOverride = marketOverrides[market];
    if (!flagOverride) {
      const alias = MARKET_ALIASES[market];
      flagOverride = alias ? marketOverrides[alias] : undefined;
    }
    if (!flagOverride) {
      console.log(`  WARN: No bundles context override for market '${market}' in ${path.basename(filePath)}`);
      return false;
    }

    if (blocks[OLD_FLAG_BLOCK_ID]) {
      blocks[FLAG_BLOCK_ID] = blocks[FLAG_BLOCK_ID]
        ? deepMerge(blocks[FLAG_BLOCK_ID], flagOverride)
        : structuredClone(flagOverride);
      delete blocks[OLD_FLAG_BLOCK_ID];
      changed = true;
    } else if (blocks[FLAG_BLOCK_ID]) {
      const merged = deepMerge(blocks[FLAG_BLOCK_ID], flagOverride);
      if (JSON.stringify(merged) !== JSON.stringify(blocks[FLAG_BLOCK_ID])) {
        blocks[FLAG_BLOCK_ID] = merged;
        changed = true;
      }
    } else {
      blocks[FLAG_BLOCK_ID] = structuredClone(flagOverride);
      changed = true;
    }
  }

  if (main.block_order) {
    const paymentIdsToRemove = new Set(findPaymentBlockIds(blocks));
    paymentIdsToRemove.delete(PAYMENT_BLOCK_ID);
    const newOrder = fixBlockOrder(cleanBlockOrder(main.block_order, blocks), paymentIdsToRemove);
    if (JSON.stringify(newOrder) !== JSON.stringify(main.block_order)) {
      main.block_order = newOrder;
      changed = true;
    }
  }

  if (changed) saveJson(filePath, data);
  return changed;
}

function getContextFilePath(baseName, market) {
  const stem = baseName.replace(/\.json$/, "");
  return path.join(TEMPLATES_DIR, `${stem}.context.${market}.json`);
}

function applyNordicFlagToContextData(contextData, market) {
  const svg = getNordicFlagSvg(market);
  if (!svg) return false;

  const main = contextData.sections?.[MAIN_SECTION] || (contextData.sections = {})[MAIN_SECTION] || {};
  main.settings = main.settings || {};
  const blocks = main.blocks || (main.blocks = {});
  const flagBlock = blocks[FLAG_BLOCK_ID] || { disabled: false, settings: {} };

  flagBlock.disabled = false;
  flagBlock.settings = flagBlock.settings || {};
  if (flagBlock.settings.icon_1_svg === svg) {
    blocks[FLAG_BLOCK_ID] = flagBlock;
    return false;
  }

  flagBlock.settings.icon_1_svg = svg;
  blocks[FLAG_BLOCK_ID] = flagBlock;
  return true;
}

function syncNordicFlagContexts(baseName) {
  const basePath = path.join(TEMPLATES_DIR, baseName);
  const baseData = loadJson(basePath);
  if (!baseData.sections?.[MAIN_SECTION]?.blocks?.[FLAG_BLOCK_ID]) {
    return [];
  }

  const changed = [];
  for (const market of NORDIC_MARKETS) {
    const contextPath = getContextFilePath(baseName, market);
    let contextData;

    if (fs.existsSync(contextPath)) {
      contextData = loadJson(contextPath);
    } else {
      contextData = {
        context: { market },
        parent: baseName,
        sections: {
          [MAIN_SECTION]: {
            settings: {},
            blocks: {},
          },
        },
      };
    }

    contextData.context = contextData.context || {};
    contextData.context.market = market;
    contextData.parent = baseName;

    if (applyNordicFlagToContextData(contextData, market)) {
      saveJson(contextPath, contextData);
      changed.push(path.basename(contextPath));
    }
  }

  return changed;
}

function main() {
  const { flagBlock, paymentBlock } = loadBundlesReference();
  const marketOverrides = loadMarketFlagOverrides();

  const baseChanged = [];
  for (const name of fs.readdirSync(TEMPLATES_DIR).sort()) {
    if (!name.startsWith("product") || !name.endsWith(".json") || name.includes(".context.")) {
      continue;
    }
    const filePath = path.join(TEMPLATES_DIR, name);
    if (processBaseTemplate(filePath, flagBlock, paymentBlock)) {
      baseChanged.push(name);
    }
  }

  const contextChanged = [];
  for (const name of fs.readdirSync(TEMPLATES_DIR).sort()) {
    if (!name.startsWith("product") || !name.includes(".context.") || !name.endsWith(".json")) {
      continue;
    }
    const filePath = path.join(TEMPLATES_DIR, name);
    if (processContextTemplate(filePath, marketOverrides)) {
      contextChanged.push(name);
    }
  }

  const nordicChanged = [];
  for (const name of fs.readdirSync(TEMPLATES_DIR).sort()) {
    if (!name.startsWith("product") || !name.endsWith(".json") || name.includes(".context.")) {
      continue;
    }
    nordicChanged.push(...syncNordicFlagContexts(name));
  }

  console.log(`Updated ${baseChanged.length} base templates:`);
  baseChanged.forEach((name) => console.log(`  - ${name}`));
  console.log(`Updated ${contextChanged.length} context templates:`);
  contextChanged.forEach((name) => console.log(`  - ${name}`));
  console.log(`Ensured Nordic flags on ${nordicChanged.length} market contexts:`);
  nordicChanged.forEach((name) => console.log(`  - ${name}`));
}

main();

File: scripts/sync_product_highlights.py
#!/usr/bin/env python3
"""Sync product_flag and payment methods blocks from product.bundles.json to all product templates."""

import copy
import json
import re
from pathlib import Path

TEMPLATES_DIR = Path(__file__).resolve().parent.parent / "templates"
BUNDLES_FILE = TEMPLATES_DIR / "product.bundles.json"
FLAG_BLOCK_ID = "product_flag_qn3MLq"
PAYMENT_BLOCK_ID = "custom_liquid_kj8teT"
OLD_FLAG_BLOCK_ID = "product_flag_T9zT3n"
MAIN_SECTION = "main"

MARKET_ALIASES = {
    "denmark": "danmark",
    "norway": "norge",
    "sweden": "svergie",
}


def strip_json_comments(text: str) -> str:
    return re.sub(r"/\*[\s\S]*?\*/", "", text)


def load_json(path: Path) -> dict:
    return json.loads(strip_json_comments(path.read_text(encoding="utf-8")))


def save_json(path: Path, data: dict) -> None:
    original = path.read_text(encoding="utf-8")
    header_match = re.match(r"^/\*[\s\S]*?\*/\s*", original)
    header = header_match.group(0) if header_match else ""
    body = json.dumps(data, indent=2, ensure_ascii=False) + "\n"
    path.write_text(header + body, encoding="utf-8")


def deep_merge(base: dict, override: dict) -> dict:
    result = copy.deepcopy(base)
    for key, value in override.items():
        if key in result and isinstance(result[key], dict) and isinstance(value, dict):
            result[key] = deep_merge(result[key], value)
        else:
            result[key] = copy.deepcopy(value)
    return result


def find_payment_block_ids(blocks: dict) -> list[str]:
    ids = []
    for block_id, block in blocks.items():
        if block.get("type") == "custom_liquid" and block.get("name") == "Payment Methods":
            ids.append(block_id)
        if block.get("type") == "section_vso_theme_sections_payment_icons":
            ids.append(block_id)
    return ids


def fix_block_order(block_order: list[str], payment_ids_to_remove: set[str]) -> list[str]:
    order = [
        bid
        for bid in block_order
        if bid not in {OLD_FLAG_BLOCK_ID, *payment_ids_to_remove}
    ]

    if "buy_buttons" not in order:
        return order

    buy_idx = order.index("buy_buttons")
    insert_at = buy_idx + 1

    if FLAG_BLOCK_ID in order:
        order.remove(FLAG_BLOCK_ID)
    if PAYMENT_BLOCK_ID in order:
        order.remove(PAYMENT_BLOCK_ID)

    order.insert(insert_at, FLAG_BLOCK_ID)
    order.insert(insert_at + 1, PAYMENT_BLOCK_ID)
    return order


def load_bundles_reference() -> tuple[dict, dict]:
    bundles = load_json(BUNDLES_FILE)
    main = bundles["sections"][MAIN_SECTION]
    flag_block = copy.deepcopy(main["blocks"][FLAG_BLOCK_ID])
    payment_block = copy.deepcopy(main["blocks"][PAYMENT_BLOCK_ID])
    return flag_block, payment_block


def load_market_flag_overrides() -> dict[str, dict]:
    overrides: dict[str, dict] = {}
    for path in sorted(TEMPLATES_DIR.glob("product.bundles.context.*.json")):
        data = load_json(path)
        market = data.get("context", {}).get("market")
        if not market:
            continue
        block = (
            data.get("sections", {})
            .get(MAIN_SECTION, {})
            .get("blocks", {})
            .get(FLAG_BLOCK_ID)
        )
        if block:
            overrides[market] = copy.deepcopy(block)
            alias = MARKET_ALIASES.get(market)
            if alias and alias not in overrides:
                overrides[alias] = copy.deepcopy(block)
    return overrides


def process_base_template(path: Path, flag_block: dict, payment_block: dict) -> bool:
    if path.name == BUNDLES_FILE.name:
        return False

    data = load_json(path)
    main = data.get("sections", {}).get(MAIN_SECTION)
    if not main:
        return False

    blocks = main.setdefault("blocks", {})
    changed = False

    if OLD_FLAG_BLOCK_ID in blocks:
        del blocks[OLD_FLAG_BLOCK_ID]
        changed = True

    payment_ids_to_remove = set(find_payment_block_ids(blocks))
    payment_ids_to_remove.discard(PAYMENT_BLOCK_ID)

    for payment_id in payment_ids_to_remove:
        if payment_id in blocks:
            del blocks[payment_id]
            changed = True

    if blocks.get(FLAG_BLOCK_ID) != flag_block:
        blocks[FLAG_BLOCK_ID] = copy.deepcopy(flag_block)
        changed = True

    if blocks.get(PAYMENT_BLOCK_ID) != payment_block:
        blocks[PAYMENT_BLOCK_ID] = copy.deepcopy(payment_block)
        changed = True

    if "block_order" in main:
        new_order = fix_block_order(main["block_order"], payment_ids_to_remove)
        if new_order != main["block_order"]:
            main["block_order"] = new_order
            changed = True

    if changed:
        save_json(path, data)
    return changed


def process_context_template(path: Path, market_overrides: dict[str, dict]) -> bool:
    if "product.bundles.context." in path.name:
        return False

    data = load_json(path)
    market = data.get("context", {}).get("market")
    if not market:
        return False

    flag_override = market_overrides.get(market)
    if not flag_override:
        alias = MARKET_ALIASES.get(market)
        flag_override = market_overrides.get(alias) if alias else None
    if not flag_override:
        print(f"  WARN: No bundles context override for market '{market}' in {path.name}")
        return False

    main = data.get("sections", {}).get(MAIN_SECTION)
    if not main:
        return False

    blocks = main.setdefault("blocks", {})
    changed = False

    if OLD_FLAG_BLOCK_ID in blocks:
        if FLAG_BLOCK_ID not in blocks:
            blocks[FLAG_BLOCK_ID] = copy.deepcopy(flag_override)
        else:
            blocks[FLAG_BLOCK_ID] = deep_merge(blocks[FLAG_BLOCK_ID], flag_override)
        del blocks[OLD_FLAG_BLOCK_ID]
        changed = True
    elif FLAG_BLOCK_ID in blocks:
        merged = deep_merge(blocks[FLAG_BLOCK_ID], flag_override)
        if merged != blocks[FLAG_BLOCK_ID]:
            blocks[FLAG_BLOCK_ID] = merged
            changed = True
    else:
        blocks[FLAG_BLOCK_ID] = copy.deepcopy(flag_override)
        changed = True

    if "block_order" in main:
        payment_ids_to_remove = set(find_payment_block_ids(blocks))
        payment_ids_to_remove.discard(PAYMENT_BLOCK_ID)
        new_order = fix_block_order(main["block_order"], payment_ids_to_remove)
        if new_order != main["block_order"]:
            main["block_order"] = new_order
            changed = True

    if changed:
        save_json(path, data)
    return changed


def main() -> None:
    flag_block, payment_block = load_bundles_reference()
    market_overrides = load_market_flag_overrides()

    base_changed = []
    for path in sorted(TEMPLATES_DIR.glob("product*.json")):
        if ".context." in path.name:
            continue
        if process_base_template(path, flag_block, payment_block):
            base_changed.append(path.name)

    context_changed = []
    for path in sorted(TEMPLATES_DIR.glob("product*.context.*.json")):
        if process_context_template(path, market_overrides):
            context_changed.append(path.name)

    print(f"Updated {len(base_changed)} base templates:")
    for name in base_changed:
        print(f"  - {name}")
    print(f"Updated {len(context_changed)} context templates:")
    for name in context_changed:
        print(f"  - {name}")


if __name__ == "__main__":
    main()
